File: Graphs/1971. Find if Path Exists in Graph Solved/1971.cpp
#pragma GCC optimize('03','unroll-loops')

class Solution {
public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        if(source==destination) return true;

        unordered_map<int,vector<int>>adjList;
        for(auto& edge : edges){
            adjList[edge[0]].push_back(edge[1]);
            adjList[edge[1]].push_back(edge[0]);
        }

        deque<int>q;
        unordered_set<int>visited;
        visited.insert(source);
        q.push_back(source);

        while(!q.empty()){
            int curr=q.front();
            q.pop_front();
            for(int nei : adjList[curr]){
                if(nei==destination){
                    return true;
                }
                if(visited.find(nei)==visited.end()){
                    q.push_back(nei);
                    visited.insert(nei);
                }
            }
        }
        return false;
    }
};

File: Graphs/1971. Find if Path Exists in Graph Solved/1971.py
# #benjamin420
# class Solution:
#     def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
#         if source==destination:
#             return True

#         adj_list=defaultdict(list)
#         for u,v in edges:
#             adj_list[u].append(v)
#             adj_list[v].append(u)
        
#         q=deque([source])
#         visited=set([source])
#         while q:
#             curr=q.popleft()
#             for nei in adj_list[curr]:
#                 if nei==destination:
#                     return True
#                 if nei not in visited:
#                     q.append(nei)
#                     visited.add(nei)

#         return False


class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        par=[i for i in range(n)]
        rank=[1] * n

        def find(node):
            p=par[node]
            while p!=par[p]:
                par[p]=par[par[p]]
                p=par[p]
            return p
        
        def union(n1,n2):
            p1,p2=find(n1),find(n2)
            if p1==p2:
                return True
            if rank[p1]>rank[p2]:
                par[p2]=p1
                rank[p1]+=rank[p2]
            else:
                par[p1]=p2
                rank[p2]+=rank[p1]

        for n1,n2 in edges:
            union(n1,n2)
            
        return find(source)==find(destination)